### PR TITLE
Feature/nucleo wifi bsp io layer robustness

### PR DESCRIPTION
**Pull request by milgra**
[DESCRIPTION]
- added a xi_data_desc_t based buffer chain to the io layer to store incoming data chunks

[ JIRA ]
https://jira.3amlabs.net/browse/XCL-2851

[REVIEWERS]

[QA]
travis tests pass
tested with 1600 byte length packets from the CPM web mqtt client

[RELEASE NOTES]
STM32F4 Nucleo BSP robustness improvements

### CODE REVIEW

File: src/bsp/platform/stm32f4_nucleo_wifi/xi_bsp_io_net_stm32f4_nucleo_wifi.c [hunk 1]
```diff
@@ -17,22 +18,22 @@ extern "C" {
...
-typedef struct _xi_bsp_stm_wifi_buffer_t
+typedef struct _xi_bsp_stm_net_state_t
 {
-    uint8_t bytes[XI_BSP_IO_NET_BUFFER_SIZE];
-    uint32_t size;
-} xi_bsp_stm_wifi_buffer_t;
+    xi_data_desc_t* head;
+    xi_bsp_stm_wifi_state_t wifi_state;
+
+} xi_bsp_stm_net_state_t;
```
Comment (atigyi) [inline src/bsp/platform/stm32f4_nucleo_wifi/xi_bsp_io_net_stm32f4_nucleo_wifi.c]: :+1: good abstraction putting these two under one common name!

File: src/bsp/platform/stm32f4_nucleo_wifi/xi_bsp_io_net_stm32f4_nucleo_wifi.c [hunk 2]
```diff
@@ -107,28 +108,38 @@ xi_bsp_io_net_state_t xi_bsp_io_net_read( xi_bsp_socket_t xi_socket,
 
     *out_read_count = 0;
 
-    if ( xi_wifi_state == xi_wifi_state_connected )
+    if ( xi_net_state.wifi_state == xi_wifi_state_connected )
     {
-        if ( xi_wifi_buffer.size == 0 )
+        if ( xi_net_state.head == NULL )
```
Comment (atigyi) [inline src/bsp/platform/stm32f4_nucleo_wifi/xi_bsp_io_net_stm32f4_nucleo_wifi.c]: we usually do this reverse NULL == xi_net_state.head no lower the chance of setting value of the variable accidentally.

Comment (milgra) [inline src/bsp/platform/stm32f4_nucleo_wifi/xi_bsp_io_net_stm32f4_nucleo_wifi.c]: thx, updated

File: src/bsp/platform/stm32f4_nucleo_wifi/xi_bsp_io_net_stm32f4_nucleo_wifi.c [hunk 3]
```diff
@@ -107,28 +108,38 @@ xi_bsp_io_net_state_t xi_bsp_io_net_read( xi_bsp_socket_t xi_socket,
...
-        }
         else
         {
-            memcpy( buf, xi_wifi_buffer.bytes, xi_wifi_buffer.size );
-            *out_read_count     = xi_wifi_buffer.size;
-            xi_wifi_buffer.size = 0;
+            xi_data_desc_t* head = xi_net_state.head;
+
+            size_t bytes_available = head->length - head->curr_pos;
+            size_t bytes_to_copy = bytes_available > count ? count : bytes_available;
```
Comment (atigyi) [inline src/bsp/platform/stm32f4_nucleo_wifi/xi_bsp_io_net_stm32f4_nucleo_wifi.c]: two things:
1. these both can be const
2. the ? : operator used withouth parenthesis tricked me multiple times
this line
size_t bytes_to_copy = bytes_available > count ? count : bytes_available;
can mean even this:
size_t bytes_to_copy = bytes_available > (count ? count : bytes_available);
so please use the following:
size_t bytes_to_copy = (bytes_available > count) ? count : bytes_available;
thx

Comment (milgra) [inline src/bsp/platform/stm32f4_nucleo_wifi/xi_bsp_io_net_stm32f4_nucleo_wifi.c]: 👍 changing...

Comment (atigyi) [inline src/bsp/platform/stm32f4_nucleo_wifi/xi_bsp_io_net_stm32f4_nucleo_wifi.c]: or even you could use a macro for selecting the minimum of two integrals:
#define MIN( a, b ) ( ( ( a ) < ( b ) ) ? ( a ) : ( b ) )

File: src/bsp/platform/stm32f4_nucleo_wifi/xi_bsp_io_net_stm32f4_nucleo_wifi.c [hunk 4]
```diff
@@ -107,28 +108,38 @@ xi_bsp_io_net_state_t xi_bsp_io_net_read( xi_bsp_socket_t xi_socket,
...
+
+            memcpy( buf, head->data_ptr + head->curr_pos , bytes_to_copy );
+
+            head->curr_pos += bytes_to_copy;
+            *out_read_count = bytes_to_copy;
+
+            if ( head->curr_pos == head->length ) 
+            {
+                xi_data_desc_t* temp = head;
```
Comment (atigyi) [inline src/bsp/platform/stm32f4_nucleo_wifi/xi_bsp_io_net_stm32f4_nucleo_wifi.c]: you don't need the temp here

File: src/bsp/platform/stm32f4_nucleo_wifi/xi_bsp_io_net_stm32f4_nucleo_wifi.c [hunk 5]
```diff
@@ -198,28 +219,21 @@ void xi_bsp_io_net_socket_data_received_proxy( uint8_t socket_id,
...
-        }
-        else
-        {
-            memcpy( xi_wifi_buffer.bytes, data_ptr, message_size );
-            xi_wifi_buffer.size = chunk_size;
-        }
-    }
+    ( void )message_size;
+
+    xi_data_desc_t* tail = xi_make_desc_from_buffer_copy( data_ptr , chunk_size );
```
Comment (atigyi) [inline src/bsp/platform/stm32f4_nucleo_wifi/xi_bsp_io_net_stm32f4_nucleo_wifi.c]: extra space before coma. We should enable the clang-format on your editor.

File: src/bsp/platform/stm32f4_nucleo_wifi/xi_bsp_io_net_stm32f4_nucleo_wifi.c (hunk 4 above)
Comment (milgra) [inline src/bsp/platform/stm32f4_nucleo_wifi/xi_bsp_io_net_stm32f4_nucleo_wifi.c]: nice one!

File: src/bsp/platform/stm32f4_nucleo_wifi/xi_bsp_io_net_stm32f4_nucleo_wifi.c [hunk 6]
```diff
@@ -198,28 +219,21 @@ void xi_bsp_io_net_socket_data_received_proxy( uint8_t socket_id,
...
+
+    if ( xi_net_state.head == NULL ) xi_net_state.head = tail;
```
Comment (atigyi) [inline src/bsp/platform/stm32f4_nucleo_wifi/xi_bsp_io_net_stm32f4_nucleo_wifi.c]: pls use curly brances for all oneliners too. (clang-format would solve this as well)

File: src/bsp/platform/stm32f4_nucleo_wifi/xi_bsp_io_net_stm32f4_nucleo_wifi.c (hunk 5 above)
Comment (milgra) [inline src/bsp/platform/stm32f4_nucleo_wifi/xi_bsp_io_net_stm32f4_nucleo_wifi.c]: true, I forgot to run it, doing it

**Pull request merged**